File: src/main/java/com/bezkoder/springjwt/security/services/ProductService.java
package com.bezkoder.springjwt.security.services;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.Optional;

import jakarta.validation.Valid;

import com.bezkoder.springjwt.models.CartItem;
import com.bezkoder.springjwt.models.Category;
import com.bezkoder.springjwt.models.Product;
import com.bezkoder.springjwt.models.ShoppingCart;
import com.bezkoder.springjwt.models.User;
import com.bezkoder.springjwt.payload.request.CartItemRequest;
import com.bezkoder.springjwt.payload.request.CartRequest;
import com.bezkoder.springjwt.payload.request.CategoryRequest;
import com.bezkoder.springjwt.payload.request.ProductRequest;
import com.bezkoder.springjwt.payload.request.ShoppingCartRequest;
import com.bezkoder.springjwt.payload.response.CategoryResponse;
import com.bezkoder.springjwt.payload.response.MessageResponse;
import com.bezkoder.springjwt.payload.response.ProductResponse;
import com.bezkoder.springjwt.payload.response.ShoppingCartResponse;
import com.bezkoder.springjwt.repository.CartItemRepository;
import com.bezkoder.springjwt.repository.CategoryRepository;
import com.bezkoder.springjwt.repository.ProductRepository;
import com.bezkoder.springjwt.repository.ShoppingCartRepository;
import com.bezkoder.springjwt.repository.UserRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

@Service
public class ProductService {
  @Autowired
  ProductRepository productRepository;
  @Autowired
  CategoryRepository categoryRepository;
  @Autowired
  UserRepository userRepository;
  @Autowired
  ShoppingCartRepository shoppingcartRepository;
  @Autowired
  CartItemRepository cartItemRepository;
  @Autowired
  UserDetailsServiceImpl userDetailsServiceImpl;
  @Autowired
  EmailSenderService emailSenderService;
  public ResponseEntity<?> findProduct(ProductRequest productRequest) {
    if (productRequest.getCategory() != null && productRequest.getName() != null) {
      return findByNameContainingAndCategory(productRequest);
    } else {
      if (productRequest.getName() != null) {
        return findByNameContaining(productRequest);
      } else {
        return findAllProducts(productRequest);
      }
    }
  }

  public Pageable checkPaging(ProductRequest productRequest) {
    if (productRequest.getPagesize() > 0 && productRequest.getPagenumber() >= 0) {
      if (productRequest.getSortdirection() == "desc") {
        return PageRequest.of(productRequest.getPagenumber(), productRequest.getPagesize(),
            Sort.by(productRequest.getSort()).descending());
      } else {
        return PageRequest.of(productRequest.getPagenumber(), productRequest.getPagesize(),
            Sort.by(productRequest.getSort()).ascending());
      }
    }
    return null;
  }

  public ResponseEntity<?> findAllProducts(ProductRequest productRequest) {
    Pageable paging = checkPaging(productRequest);
    if (paging == null)
      return ResponseEntity.ok(new ProductResponse(productRepository.findAll()));
    return ResponseEntity.ok(new ProductResponse(productRepository.findAll(paging)));
  }

  public ResponseEntity<?> findByNameContaining(ProductRequest productRequest) {
    Pageable paging = checkPaging(productRequest);
    if (paging == null)
      return ResponseEntity.ok(new ProductResponse(productRepository.findByNameContaining(productRequest.getName())));
    else
      return ResponseEntity
          .ok(new ProductResponse(productRepository.findByNameContaining(productRequest.getName(), paging)));
  }

  public ResponseEntity<?> findByNameContainingAndCategory(ProductRequest productRequest) {
    Pageable paging = checkPaging(productRequest);
    if (paging == null)
      return ResponseEntity.ok(new ProductResponse(
          productRepository.findByNameContainingAndCategory(productRequest.getName(), productRequest.getCategory())));
    else
      return ResponseEntity.ok(new ProductResponse(productRepository
          .findByNameContainingAndCategory(productRequest.getName(), productRequest.getCategory(), paging)));
  }

  public ResponseEntity<?> addProduct(ProductRequest productRequest) {
    userDetailsServiceImpl.checkAdmin(productRequest.getUsername(), productRequest.getPassword());

    Category category = categoryRepository.findById(productRequest.getCategory().getId()).get();
    if (category == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: Category does not exist"));
    if (productRequest.getName() == null || productRequest.getName() == "")
      return ResponseEntity.badRequest().body(new MessageResponse("Error: name must not be empty"));
    if (productRequest.getPrice() == null || productRequest.getPrice() <= 0)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: price must not be empty, zero or negative"));
    Product product = new Product(productRequest);
    productRepository.save(product);
    return ResponseEntity.ok(new MessageResponse("product added successfully!"));
  }

  public ResponseEntity<?> updateProduct(ProductRequest productRequest) {
    userDetailsServiceImpl.checkAdmin(productRequest.getUsername(), productRequest.getPassword());

    Category category = categoryRepository.findById(productRequest.getCategory().getId()).get();
    if (category == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: Category does not exist"));
    if (productRequest.getName() == null || productRequest.getName() == "")
      return ResponseEntity.badRequest().body(new MessageResponse("Error: name must not be empty"));
    if (productRequest.getPrice() == null || productRequest.getPrice() <= 0)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: price must not be empty, zero or negative"));
    Product product = new Product(productRequest);
    productRepository.save(product);
    return ResponseEntity.ok(new MessageResponse("product updated successfully!"));
  }

  public ResponseEntity<?> deleteProduct(ProductRequest productRequest) {
    userDetailsServiceImpl.checkAdmin(productRequest.getUsername(), productRequest.getPassword());

    productRepository.deleteById(productRequest.getId());
    return ResponseEntity.ok(new MessageResponse("product deleted successfully"));
  }

  public ResponseEntity<?> saveProductImage(long id, MultipartFile image, int number) {

    boolean exists = true;
    if (number < 0 || number > 4)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: product can only have 5 images"));
    Product Product = getProduct(id);
    OutputStream out = null;
    try {
      Files.createDirectories(Paths.get("src/assets/productimages/" + id + "/"));
      File tempFile = new File("src/assets/productimages/" + id + "/" + number + ".jpg");
      exists = tempFile.exists();
      out = new FileOutputStream("src/assets/productimages/" + id + "/" + number + ".jpg");
      out.write(image.getBytes());
      Product.setImages(true);
      productRepository.save(Product);
      if (exists)
        return ResponseEntity.ok(new MessageResponse("product image updated successfully"));
      else
        return ResponseEntity.ok(new MessageResponse("product image added successfully"));
    } catch (IOException e) {
      e.printStackTrace();
      return ResponseEntity.badRequest().body(new MessageResponse("Error: image could not be saved"));
    } finally {
      try {
        out.flush();
        out.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
  }

  private Product getProduct(long id) {

    Product product = productRepository.findById(id).get();
    checkProduct(product);
    return product;
  }

  private void checkProduct(Product product) {
    if (product == null)
      productDoesNotExist();
  }

  private ResponseEntity<?> productDoesNotExist() {
    return ResponseEntity.badRequest().body(new MessageResponse("Error: product does not exist"));
  }

  public ResponseEntity<?> updateCategory(@Valid CategoryRequest categoryRequest) {
    userDetailsServiceImpl.checkAdmin(categoryRequest.getUsername(), categoryRequest.getPassword());
    Category category = categoryRepository.findById(categoryRequest.getId()).get();
    category.setName(categoryRequest.getName());
    categoryRepository.save(category);
    return ResponseEntity.ok(new MessageResponse("Category updated successfully!"));
  }

  public ResponseEntity<?> deleteCategory(@Valid CategoryRequest categoryRequest) {
    userDetailsServiceImpl.checkAdmin(categoryRequest.getUsername(), categoryRequest.getPassword());
    categoryRepository.deleteById(categoryRequest.getId());
    return ResponseEntity.ok(new MessageResponse("Category deleted successfully!"));
  }

  public ResponseEntity<?> addCategory(@Valid CategoryRequest categoryRequest) {
    userDetailsServiceImpl.checkAdmin(categoryRequest.getUsername(), categoryRequest.getPassword());
    Category category = new Category(categoryRequest.getName());
    categoryRepository.save(category);
    return ResponseEntity.ok(new MessageResponse("Category added successfully!"));
  }

  public ResponseEntity<?> getAllCategories(CategoryRequest categoryRequest) {
    userDetailsServiceImpl.checkAdmin(categoryRequest.getUsername(), categoryRequest.getPassword());

    List<Category> categorylist = categoryRepository.findAll();
    return ResponseEntity.ok(new CategoryResponse(categorylist));
  }

  public ResponseEntity<?> addToCart(CartRequest cartRequest) {
    userDetailsServiceImpl.checkAdminOrConcernedUser(cartRequest.getUsername(), cartRequest.getPassword(),
        cartRequest.getUserid());

    // if (cartRequest.getQuantity() <= 0)
    // return ResponseEntity.badRequest().body(new MessageResponse("Error: quantity
    // must not bigger than 1"));
    if (cartRequest.getProductid() <= 0)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: product not valid"));
    if (cartRequest.getUserid() == null || cartRequest.getUserid() <= 0)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: cart's user is not valid"));
    User user = null;
    Optional<User> ouser = userRepository.findById(cartRequest.getUserid());
    if (ouser.isEmpty()) {
      userDetailsServiceImpl.userNotFound();
    } else
      user = ouser.get();
    ShoppingCart shoppingcart;
    if (user.getShoppingcart() == null) {
      shoppingcart = new ShoppingCart(user);
      shoppingcartRepository.save(shoppingcart);
      user.setShoppingcart(shoppingcart);
      userRepository.save(user);
    }

    Product product = productRepository.findById(cartRequest.getProductid()).get();

    if (product == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: product does not exist"));

    CartItem cartitem = null;
    Optional<CartItem> ocartitem = cartItemRepository.findByShoppingcartAndProduct(user.getShoppingcart(), product);
    if (ocartitem.isEmpty()) {
      cartitem = new CartItem(product, cartRequest.getQuantity());
      cartitem.setShoppingcart(user.getShoppingcart());
      cartItemRepository.save(cartitem);
    } else
      cartitem = ocartitem.get();
    shoppingcart = user.getShoppingcart();
    shoppingcart.addProduct(cartitem);
    shoppingcartRepository.save(shoppingcart);
    return ResponseEntity.ok(new MessageResponse("product added to cart"));
  }

  public ResponseEntity<?> removeCartItem(@Valid CartRequest cartRequest) {
    userDetailsServiceImpl.checkAdminOrConcernedUser(cartRequest.getUsername(), cartRequest.getPassword(),
        cartRequest.getUserid());
    User user = userRepository.findByUsernameIgnoreCase(cartRequest.getUsername()).get();
    ShoppingCart shoppingCart = user.getShoppingcart();
    Product product = getProduct(cartRequest.getProductid());

    CartItem cartItem = cartItemRepository.findByShoppingcartAndProduct(shoppingCart, product).get();
    shoppingCart.removeProduct(cartItem);
    shoppingcartRepository.save(shoppingCart);
    cartItemRepository.delete(cartItem);
    return ResponseEntity.ok(new MessageResponse("Item removed"));

  }

  public ResponseEntity<?> getShoppingcart(CartRequest cartRequest) {
    userDetailsServiceImpl.checkAdminOrConcernedUser(cartRequest.getUsername(), cartRequest.getPassword(),
        cartRequest.getUserid());
    if (cartRequest.getUserid() == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: cart's user is not valid"));
    User user = userRepository.findById(cartRequest.getUserid()).get();
    if (user == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: user does not exist"));
    ShoppingCart shoppingcart = user.getShoppingcart();
    if (shoppingcart == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: user has no cart"));
    if (shoppingcart.getCartItems().isEmpty())
      return ResponseEntity.badRequest().body(new MessageResponse("Error: empty cart"));
    return ResponseEntity.ok(new ShoppingCartResponse(shoppingcart));
  }

  public ResponseEntity<?> getAllShoppingCarts(CartRequest cartRequest) {
    userDetailsServiceImpl.checkAdmin(cartRequest.getUsername(), cartRequest.getPassword());
    List<ShoppingCart> shoppingcart = shoppingcartRepository.findByCartItemsIsNotEmpty();
    if (shoppingcart == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: no carts"));
    return ResponseEntity.ok(new ShoppingCartResponse(shoppingcart));
  }

  public ResponseEntity<?> setShippingDate(@Valid ShoppingCartRequest shoppingcartRequest) {
    userDetailsServiceImpl.checkAdmin(shoppingcartRequest.getUsername(), shoppingcartRequest.getPassword());
    if (shoppingcartRequest.getUserid() == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: cart's user is not valid"));
    User user = userRepository.findById(shoppingcartRequest.getUserid()).get();
    if (user == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: user does not exist"));
    ShoppingCart shoppingcart = shoppingcartRepository.findByUser(user).get();
    if (shoppingcart == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: user has no cart"));
    if (shoppingcart.getCartItems().isEmpty())
      return ResponseEntity.badRequest().body(new MessageResponse("Error: empty cart"));
    if (shoppingcart.getShippingdate() != null)
      if (shoppingcartRequest.getShippingdate().before(shoppingcart.getShippingdate()))
        return ResponseEntity.badRequest().body(new MessageResponse("Error: invalid date"));
    if (shoppingcartRequest.getShippingdate() == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: invalid date"));
    shoppingcart.setShippingdate(shoppingcartRequest.getShippingdate());
    shoppingcartRepository.save(shoppingcart);
    //send email
    SimpleMailMessage mailMessage = new SimpleMailMessage();
    mailMessage.setTo(user.getEmail());
    mailMessage.setSubject("Your order has been shipped!");
    mailMessage.setFrom("mrissaoussama@gmail.com");
    mailMessage.setText("hello "+user.getUsername()+", your order has been shipped");
    emailSenderService.sendEmail(mailMessage);
    return ResponseEntity.ok(new ShoppingCartResponse(shoppingcart));
  }

  public ResponseEntity<?> setCompletedDate(@Valid ShoppingCartRequest shoppingcartRequest) {
    userDetailsServiceImpl.checkAdmin(shoppingcartRequest.getUsername(), shoppingcartRequest.getPassword());
    if (shoppingcartRequest.getUserid() == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: cart's user is not valid"));
    User user = userRepository.findById(shoppingcartRequest.getUserid()).get();
    if (user == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: user does not exist"));
    ShoppingCart shoppingcart = shoppingcartRepository.findByUser(user).get();
    if (shoppingcart == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: user has no cart"));
    if (shoppingcart.getCartItems().isEmpty())
      return ResponseEntity.badRequest().body(new MessageResponse("Error: empty cart"));
    if (shoppingcart.getCompleteddate() != null)
      if (shoppingcartRequest.getCompleteddate().before(shoppingcart.getCompleteddate()))
        return ResponseEntity.badRequest().body(new MessageResponse("Error: invalid date"));
    if (shoppingcartRequest.getCompleteddate() == null)
      return ResponseEntity.badRequest().body(new MessageResponse("Error: invalid date"));
    shoppingcart.setCompleteddate(shoppingcartRequest.getCompleteddate());
    shoppingcartRepository.save(shoppingcart);
     //send email
     SimpleMailMessage mailMessage = new SimpleMailMessage();
     mailMessage.setTo(user.getEmail());
     mailMessage.setSubject("Your order has been completed!");
     mailMessage.setFrom("mrissaoussama@gmail.com");
     mailMessage.setText("hello "+user.getUsername()+", your order has been completed");
     emailSenderService.sendEmail(mailMessage);
    return ResponseEntity.ok(new ShoppingCartResponse(shoppingcart));
  }

  public ResponseEntity<?> updateShoppingCartItems(@Valid CartItemRequest cartItemRequest) {
    userDetailsServiceImpl.checkAdminOrConcernedUser(cartItemRequest.getUsername(), cartItemRequest.getPassword(),
        cartItemRequest.getUserid());

    if (cartItemRequest.getCartItems().isEmpty())
      return ResponseEntity.badRequest().body(new MessageResponse("Error: empty cart"));
    User user = userRepository.findById(cartItemRequest.getUserid()).get();

    ShoppingCart shoppingcart = user.getShoppingcart();
    shoppingcart.updateProduct(cartItemRequest.getCartItems());
    shoppingcartRepository.save(shoppingcart);
    return ResponseEntity.ok(new MessageResponse("Cart updated"));

  }

}

File: src/app/_services/product.service.ts
import { Cartitem } from './../models/cartitem.model';
import { TokenStorageService } from 'src/app/_services/token-storage.service';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { catchError, map, Observable, of } from 'rxjs';
import { Category } from './../models/category.model';
import { Product } from './../models/product.model';
import { ProductRequest } from './../models/productrequest.model';
import { NumberSymbol } from '@angular/common';

const PRODUCT_API = 'http://localhost:8080/api/product/';

const httpOptions = {
  headers: new HttpHeaders({ 'Content-Type': 'application/json' })
};
const HttpUploadOptions = {
  headers: new HttpHeaders({})
}

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  constructor(private http: HttpClient, private token: TokenStorageService) { }

  fileExists(id: any, number: any): Observable<boolean> {
    return this.http.get("/assets/Productimages/" + id + "/" + number + ".jpg").pipe(map(() => true),
      catchError(() => of(false)));
  }
  productrequest: ProductRequest = new ProductRequest('', '', 0, '', '', 0, null, false);

  updateProduct(product: Product): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();
    this.productrequest.convert(username, password, product);
    var username = this.productrequest.username;
    var password = this.productrequest.password;
    var id = this.productrequest.id;
    var category = this.productrequest.category;
    var name = this.productrequest.name;
    var price = this.productrequest.price;
    var description = this.productrequest.description;

    return this.http.post(PRODUCT_API + 'updateProduct', {//productrequest
      id, username, password, category, price, name, description

    }, httpOptions);
  }

  addProduct(form: FormGroup): Observable<any> {
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    var category = form.get('Category').value;
    var name = form.get('name').value;
    var price = form.get('price').value;
    var description = form.get('description').value;
    return this.http.post(PRODUCT_API + 'addProduct', {
      username, password, category, price,
      name
      , description
    }, httpOptions);
  }

  updateProductImage(form: FormData): Observable<any> {
    return this.http.post(PRODUCT_API + 'updateProductImage', form);
  }

  findProduct(productrequest: ProductRequest): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();
    var id = productrequest.id;
    var category = productrequest.category;
    var name = productrequest.name;
    var price = productrequest.price;
    var description = productrequest.description;
    console.log(productrequest)

    return this.http.post(PRODUCT_API + 'findProduct', {

      id, username, password, category, price, name, description
    }, httpOptions);
  }
  deleteProduct(product: Product): Observable<any> {
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    var id = product.id
    return this.http.post(PRODUCT_API + 'deleteProduct', {

      username,
      id,
      password
    }, httpOptions);
  }
  getAllProducts(): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();
    return this.http.post(PRODUCT_API + 'findProduct', {
      username, password
    }, httpOptions);
  }
  getAllCategories(): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();
    return this.http.post(PRODUCT_API + 'getAllCategories', {
      username, password
    }, httpOptions);
  }
  updateCategory(category: Category): Observable<any> {
    var id = category.id;
    var name = category.name
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    // var price=form.get('price').value;
    // var description=form.get('description').value;
    return this.http.post(PRODUCT_API + 'updateCategory', {
      id, name,
      username, password
    }, httpOptions);
  }
  addCategory(category: Category): Observable<any> {
    var name = category.name
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    // var price=form.get('price').value;
    // var description=form.get('description').value;
    return this.http.post(PRODUCT_API + 'addCategory', {
      name,
      username, password
    }, httpOptions);
  }
  deleteCategory(category: Category): Observable<any> {
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    var id = category.id
    var name = category.name;
    return this.http.post(PRODUCT_API + 'deleteCategory', {
      username,
      id,
      name,
      password
    }, httpOptions);
  }
  getAllShoppingCarts(): Observable<any> {
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    return this.http.post(PRODUCT_API + 'getAllShoppingCarts', {
      username, password
    }, httpOptions);
  }
  getShoppingCart(userid: any): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();

    return this.http.post(PRODUCT_API + 'getShoppingCart', {
      username, password, userid
    }, httpOptions);
  }
  updateShoppingCartItems(userid: any, cartItems: Cartitem[]): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();
    cartItems.forEach(element => {
      element.shoppingcart = undefined
    });
    return this.http.post(PRODUCT_API + 'updateShoppingCartItems', {
      username, password, userid, cartItems
    }, httpOptions);
  }
  removeCartItem(userid: any, productid: number): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();

    return this.http.post(PRODUCT_API + 'removeCartItem', {
      username, password, userid, productid
    }, httpOptions);
  }
  setCompletedDate(shoppingcart: any): Observable<any> {

    var username = this.token.getUsername();
    var password = this.token.getPassword();
    var userid = shoppingcart.user.id;
    // var price=form.get('price').value;
    // var description=form.get('description').value;
    return this.http.post(PRODUCT_API + 'setCompletedDate', {
      username, password, userid
    }, httpOptions);
  }
  setShippingDate(shoppingcart: any): Observable<any> {
    var username = this.token.getUsername();
    var password = this.token.getPassword();
    var userid = shoppingcart.user.id;
    // var price=form.get('price').value;
    // var description=form.get('description').value;
    return this.http.post(PRODUCT_API + 'setShippingDate', {
      username, password, userid
    }, httpOptions);
  }
}
